<img style="width:30%; float:right" src="images/dbu_logo.png">

# Manipulation von Daten

Wir können in Pandas mit einfachen Befehlen Spalten und Zeilen anlegen, löschen oder verändern.

**Literaturempfehlung:**

[Jake VanderPlas (2016): _Python Data Science Handbook_](https://ebookcentral.proquest.com/lib/dbuas/reader.action?docID=4746657)

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
tips = sns.load_dataset("tips")
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## Neue Spalten anlegen und Spalten löschen

In [3]:
tips["nonsense"] = "abc"
tips.head()

total_bill   tip     sex smoker  day    time  size nonsense
0       16.99  1.01  Female     No  Sun  Dinner     2      abc
1       10.34  1.66    Male     No  Sun  Dinner     3      abc
2       21.01  3.50    Male     No  Sun  Dinner     3      abc
3       23.68  3.31    Male     No  Sun  Dinner     2      abc
4       24.59  3.61  Female     No  Sun  Dinner     4      abc

In [4]:
tips["nonsense"] = "xyz"
tips.head()

total_bill   tip     sex smoker  day    time  size nonsense
0       16.99  1.01  Female     No  Sun  Dinner     2      xyz
1       10.34  1.66    Male     No  Sun  Dinner     3      xyz
2       21.01  3.50    Male     No  Sun  Dinner     3      xyz
3       23.68  3.31    Male     No  Sun  Dinner     2      xyz
4       24.59  3.61  Female     No  Sun  Dinner     4      xyz

In [5]:
tips.drop("nonsense", axis="columns", inplace=True)
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## Neue Spalten berechnen

In [6]:
tips["tip_percent"] = tips["tip"] / tips["total_bill"]
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percent
0       16.99  1.01  Female     No  Sun  Dinner     2     0.059447
1       10.34  1.66    Male     No  Sun  Dinner     3     0.160542
2       21.01  3.50    Male     No  Sun  Dinner     3     0.166587
3       23.68  3.31    Male     No  Sun  Dinner     2     0.139780
4       24.59  3.61  Female     No  Sun  Dinner     4     0.146808

In [7]:
tips["total_in_euro"] = tips["total_bill"] * 0.91
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percent  \
0       16.99  1.01  Female     No  Sun  Dinner     2     0.059447   
1       10.34  1.66    Male     No  Sun  Dinner     3     0.160542   
2       21.01  3.50    Male     No  Sun  Dinner     3     0.166587   
3       23.68  3.31    Male     No  Sun  Dinner     2     0.139780   
4       24.59  3.61  Female     No  Sun  Dinner     4     0.146808   

   total_in_euro  
0        15.4609  
1         9.4094  
2        19.1191  
3        21.5488  
4        22.3769

In [8]:
tips["total_in_euro"] = tips["total_in_euro"].round(2)
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percent  \
0       16.99  1.01  Female     No  Sun  Dinner     2     0.059447   
1       10.34  1.66    Male     No  Sun  Dinner     3     0.160542   
2       21.01  3.50    Male     No  Sun  Dinner     3     0.166587   
3       23.68  3.31    Male     No  Sun  Dinner     2     0.139780   
4       24.59  3.61  Female     No  Sun  Dinner     4     0.146808   

   total_in_euro  
0          15.46  
1           9.41  
2          19.12  
3          21.55  
4          22.38

## Neue Zeile anlegen

In [9]:
new_row = pd.Series([5, 1, "Sun"], index=["total_bill", "tip", "day"])
new_row

total_bill      5
tip             1
day           Sun
dtype: object

In [10]:
tips = tips.append(new_row, ignore_index=True)
tips[-5:]

total_bill   tip     sex smoker   day    time  size  tip_percent  \
240       27.18  2.00  Female    Yes   Sat  Dinner   2.0     0.073584   
241       22.67  2.00    Male    Yes   Sat  Dinner   2.0     0.088222   
242       17.82  1.75    Male     No   Sat  Dinner   2.0     0.098204   
243       18.78  3.00  Female     No  Thur  Dinner   2.0     0.159744   
244        5.00  1.00     NaN    NaN   Sun     NaN   NaN          NaN   

     total_in_euro  
240          24.73  
241          20.63  
242          16.22  
243          17.09  
244            NaN

## Umgang mit "falschen" Werten

In [11]:
tips.loc[tips["tip_percent"] > 0.3]

total_bill   tip     sex smoker  day    time  size  tip_percent  \
67         3.07  1.00  Female    Yes  Sat  Dinner   1.0     0.325733   
172        7.25  5.15    Male    Yes  Sun  Dinner   2.0     0.710345   
178        9.60  4.00  Female    Yes  Sun  Dinner   2.0     0.416667   

     total_in_euro  
67            2.79  
172           6.60  
178           8.74

In [12]:
tips.loc[tips["tip_percent"] > 0.3, "tip"] = np.NaN

In [13]:
tips.loc[tips["tip_percent"] > 0.3]

total_bill  tip     sex smoker  day    time  size  tip_percent  \
67         3.07  NaN  Female    Yes  Sat  Dinner   1.0     0.325733   
172        7.25  NaN    Male    Yes  Sun  Dinner   2.0     0.710345   
178        9.60  NaN  Female    Yes  Sun  Dinner   2.0     0.416667   

     total_in_euro  
67            2.79  
172           6.60  
178           8.74

# Aufgaben

## Aufgabe 1:

Wir interessieren uns dafür, wieviel bei einer Mahlzeit pro Person ausgegeben wurde.

1. Erstelle dazu einen neue Spalte "total_per_person".
2. Wie bei den Euro-Beträgen, sollte die Spalte auf genau zwei Nachkommastellen genau sein.

## Aufgabe 2:

Die Spalte "tip_percent" stellt genau genommen noch keine echten Prozente dar, sondern Anteile zwischen 0 und 1 statt zwischen 0 und 100. Wo jetzt `0.059447` steht hätten wir also gerne `5.94` (uns reichen für den Momenten zwei Nachkommastellen).

Überschreibe die Spalte "tip_precent" entsprechend, dass beispielsweise `0.059447` zu `5.94` wird.